# Exercicis amb Dataframes.

## - Exercici 1
### Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

#### ->Airlines Delay: Airline on-time statistics and delay causes


In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

In [3]:
df = pd.read_csv('DelayedFlights.csv')
df.head(10)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   
5           6  2008      1           3          4   1937.0        1830   
6          10  2008      1           3          4    706.0         700   
7          11  2008      1           3          4   1644.0        1510   
8          15  2008      1           3          4   1029.0        1020   
9          16  2008      1           3          4   1452.0        1425   

   ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   2211.0        2225            WN        335  N712SW              128.0   
1   1002.0        1000            WN       3231  N772SW              128.0   
2    804.0         750            WN        448  N428WN               96.0   
3   1959.0        1925            WN       3920  N464WN               90.0   
4   2121.0        2110            WN        378  N726SW              101.0   
5   2037.0        1940            WN        509  N763SW              240.0   
6    916.0         915            WN        100  N690SW              130.0   
7   1845.0        1725            WN       1333  N334SW              121.0   
8   1021.0        1010            WN       2272  N263WN               52.0   
9   1640.0        1625            WN        675  N286WN              228.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0           150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1           145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2            90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
3            90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
4           115.0     87.0      11.0      25.0    IND  JAX       688     4.0   
5           250.0    230.0      57.0      67.0    IND  LAS      1591     3.0   
6           135.0    106.0       1.0       6.0    IND  MCO       828     5.0   
7           135.0    107.0      80.0      94.0    IND  MCO       828     6.0   
8            50.0     37.0      11.0       9.0    IND  MDW       162     6.0   
9           240.0    213.0      15.0      27.0    IND  PHX      1489     7.0   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0      8.0          0                N         0           NaN           NaN   
1     10.0          0                N         0           NaN           NaN   
2     17.0          0                N         0           NaN           NaN   
3     10.0          0                N         0           2.0           0.0   
4     10.0          0                N         0           NaN           NaN   
5      7.0          0                N         0          10.0           0.0   
6     19.0          0                N         0           NaN           NaN   
7      8.0          0                N         0           8.0           0.0   
8      9.0          0                N         0           NaN           NaN   
9      8.0          0                N         0           3.0           0.0   

   NASDelay  SecurityDelay  LateAircraftDelay  
0       NaN            NaN                NaN  
1       NaN            NaN                NaN  
2       NaN            NaN                NaN  
3       0.0            0.0               32.0  
4       NaN            NaN                NaN  
5       0.0            0.0               47.0  
6       NaN            NaN                NaN  
7       0.0            0.0               72.0  
8       NaN            NaN                NaN  
9       0.0            0.0               12.0

### Existe 1.936.758 registros en donde consta la informacion de 30 variables representadas en columnas las cuales son:

### Year 2008 
### Month 1-12
### DayofMonth 1-31
### DayOfWeek 1 (Monday) - 7 (Sunday) 
### DepTime actual departure time (local, hhmm) (hora de despegue)
### CRSDepTime scheduled departure time (local, hhmm) (hora de despegue programada)
### ArrTime actual arrival time (local, hhmm) (hora de llegada)
### CRSArrTime scheduled arrival time (local, hhmm) (hora de llegada estimada)
### UniqueCarrier unique carrier code (codigo de la aerolinea)
### FlightNum flight number
### TailNum plane tail number: aircraft registration, unique aircraft identifier 
### ActualElapsedTime in minutes (tiempo transcurrido real)
### CRSElapsedTime in minutes (tiempo transcurrido estimado)
### AirTime in minutes (tiempo de vuelo)
### ArrDelay arrival delay, in minutes: A flight is counted as "on time" if it operated less than 15 minutes later the scheduled time shown in the carriers' Computerized Reservations Systems (CRS). (los vuelos son considerados llegados a tiempo cuando el vuelo tiene un retraso inferior a 15 minutos) 
### DepDelay departure delay, in minutes
### Origin origin IATA airport code
### Dest destination IATA airport code
### Distance in miles
### TaxiIn taxi in time, in minutes (tiempo de rodaje en pista)
### TaxiOut taxi out time in minutes (tiempo de rodaje en pista)
### Cancelled *was the flight cancelled
### CancellationCode reason for cancellation (A = carrier, B = weather, C = NAS, D = security)
### Diverted 1 = yes, 0 = no
### CarrierDelay in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.
### WeatherDelay in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.
### NASDelay in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.
### SecurityDelay in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.
### LateAircraftDelay in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagation.


## queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

### Una vez que conocemos que representa cada columna podemos pasar a limpiar el Dataset y ver que informacion no sera relevante para el analisis que deseamos hacer, para comenzar podemos ver que las columnas referidas a las causas de los retrasos CarrierDelay, WeatherDelay,	NASDelay, SecurityDelay,	LateAircraftDelay contienen en su mayoria valores NaN por lo que no las tendremos en cuenta y seran apartadas. Ademas considerando los ejercicios planteados consideros que las columnas de CancellationCode y Any no seran necesarias.

In [4]:
mis_columnas = ['Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay', 'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut', 'Cancelled']
my_df = df[mis_columnas]
my_df.head(21)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0       1           3          4   2003.0        1955   2211.0        2225   
1       1           3          4    754.0         735   1002.0        1000   
2       1           3          4    628.0         620    804.0         750   
3       1           3          4   1829.0        1755   1959.0        1925   
4       1           3          4   1940.0        1915   2121.0        2110   
5       1           3          4   1937.0        1830   2037.0        1940   
6       1           3          4    706.0         700    916.0         915   
7       1           3          4   1644.0        1510   1845.0        1725   
8       1           3          4   1029.0        1020   1021.0        1010   
9       1           3          4   1452.0        1425   1640.0        1625   
10      1           3          4    754.0         745    940.0         955   
11      1           3          4   1323.0        1255   1526.0        1510   
12      1           3          4   1416.0        1325   1512.0        1435   
13      1           3          4   1657.0        1625   1754.0        1735   
14      1           3          4   1900.0        1840   1956.0        1950   
15      1           3          4   1039.0        1030   1133.0        1140   
16      1           3          4   1520.0        1455   1619.0        1605   
17      1           3          4   1422.0        1255   1657.0        1610   
18      1           3          4   1954.0        1925   2239.0        2235   
19      1           3          4   2107.0        1945   2334.0        2230   
20      1           3          4   1312.0        1300   1546.0        1550   

   UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0             WN        335  N712SW              128.0           150.0   
1             WN       3231  N772SW              128.0           145.0   
2             WN        448  N428WN               96.0            90.0   
3             WN       3920  N464WN               90.0            90.0   
4             WN        378  N726SW              101.0           115.0   
5             WN        509  N763SW              240.0           250.0   
6             WN        100  N690SW              130.0           135.0   
7             WN       1333  N334SW              121.0           135.0   
8             WN       2272  N263WN               52.0            50.0   
9             WN        675  N286WN              228.0           240.0   
10            WN       1144  N778SW              226.0           250.0   
11            WN          4  N674AA              123.0           135.0   
12            WN         54  N643SW               56.0            70.0   
13            WN        623  N724SW               57.0            70.0   
14            WN        717  N786SW               56.0            70.0   
15            WN       1244  N714CB               54.0            70.0   
16            WN       2553  N394SW               59.0            70.0   
17            WN        188  N215WN              155.0           195.0   
18            WN       1754  N243WN              165.0           190.0   
19            WN        362  N798SW              147.0           165.0   
20            WN       1397  N247WN              154.0           170.0   

    AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0     116.0     -14.0       8.0    IAD  TPA       810     4.0      8.0   
1     113.0       2.0      19.0    IAD  TPA       810     5.0     10.0   
2      76.0      14.0       8.0    IND  BWI       515     3.0     17.0   
3      77.0      34.0      34.0    IND  BWI       515     3.0     10.0   
4      87.0      11.0      25.0    IND  JAX       688     4.0     10.0   
5     230.0      57.0      67.0    IND  LAS      1591     3.0      7.0   
6     106.0       1.0       6.0    IND  MCO       828     5.0     19.0   
7     107.0      80.0      94.0    IND  MCO       828     6.0     

### considerando los ejercicicios propustos las columnas de nuestro interes seran: Month,  DayofMonth,	DayOfWeek,	DepTime,	CRSDepTime,	ArrTime,	CRSArrTime,	UniqueCarrier,	FlightNum,	TailNum,	ActualElapsedTime,	CRSElapsedTime,	AirTime	ArrDelay,	DepDelay,	Origin,	Dest,	Distance,	TaxiIn,	TaxiOut,	Cancelled.

In [6]:
my_df_sampled = my_df.sample(n=200000, random_state=42)

### En el paso anterior reducimos nuestro DataFrame a 200.000 registros de manera aleaotria

## - Exercici 2
### Fes un informe complet del dataset:

#### Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

In [62]:
my_df_sampled.describe().T

count         mean          std   min     25%     50%  \
Month              200000.0     6.106975     3.476865   1.0     3.0     6.0   
DayofMonth         200000.0    15.734955     8.770451   1.0     8.0    16.0   
DayOfWeek          200000.0     3.983700     1.995799   1.0     2.0     4.0   
DepTime            200000.0  1520.114805   450.087506   1.0  1205.0  1546.0   
CRSDepTime         200000.0  1468.567445   424.431628   0.0  1135.0  1510.0   
ArrTime            199267.0  1611.293159   548.219910   1.0  1317.0  1716.0   
CRSArrTime         200000.0  1635.514525   464.072463   0.0  1325.0  1707.0   
FlightNum          200000.0  2188.548195  1944.372086   1.0   615.0  1549.0   
ActualElapsedTime  199127.0   133.222772    72.001365  15.0    80.0   116.0   
CRSElapsedTime     199974.0   134.184889    71.177993   1.0    82.0   116.0   
AirTime            199127.0   108.197070    68.472446   0.0    58.0    90.0   
ArrDelay           199127.0    42.371456    57.004644 -57.0     9.0    25.0   
DepDelay           200000.0    43.319360    53.581396   6.0    12.0    24.0   
Distance           200000.0   764.961740   572.856710  11.0   340.0   607.0   
TaxiIn             199267.0     6.817747     5.342789   0.0     4.0     6.0   
TaxiOut            199956.0    18.217083    14.410192   0.0    10.0    14.0   
Cancelled          200000.0     0.000300     0.017318   0.0     0.0     0.0   

                      75%     max  
Month                 9.0    12.0  
DayofMonth           23.0    31.0  
DayOfWeek             6.0     7.0  
DepTime            1901.0  2400.0  
CRSDepTime         1817.0  2359.0  
ArrTime            2031.0  2400.0  
CRSArrTime         2015.0  2400.0  
FlightNum          3423.0  9740.0  
ActualElapsedTime   165.0   666.0  
CRSElapsedTime      165.0   660.0  
AirTime             137.0   644.0  
ArrDelay             56.0  2453.0  
DepDelay             54.0  2467.0  
Distance            998.0  4962.0  
TaxiIn                8.0   200.0  
TaxiOut              21.0   342.0  
Cancelled             0.0     1.0

### Con el metodo describe( ) hacemos un analisis estadistico del data set que nos devuelve nformación que incluye el número de muestras, el valor medio, la desviación estándar, el valor mínimo, máximo, la mediana y los valores correspondientes a los percentiles 25% y 75% con esta informacion podemos tener una idea mas rapida de como se distribuyen los datos y sus caracteristicas.

### Los datos mas relevantes son los de las columnas (ActualElapsedTime, AirTime,	ArrDelay,	DepDelay,	Distance,	TaxiIn,	TaxiOut) ya que nos permite tener una idea aproximada de los valores reales que se mantuvieron durante los vuelos y sus posibles retrasos.

### Tiempo máximo de vuelo (ActualElapsedTime) fue de 666 minutos.
### Tiempo minimo de vuelo fue de 15 minutos 
### Media tiempo de vuelo 133.30 minutos.

### Tiempo maximo de retraso (ArrDelay) 2453.0 minutos
### Media de tiempo de retraso 43.31 minutos

## Troba quantes dades faltants hi ha per columna.

In [73]:
DatosFaltantes = my_df_sampled.isna().sum()
print(DatosFaltantes)

Month                  0
DayofMonth             0
DayOfWeek              0
DepTime                0
CRSDepTime             0
ArrTime              733
CRSArrTime             0
UniqueCarrier          0
FlightNum              0
TailNum                0
ActualElapsedTime    873
CRSElapsedTime        26
AirTime              873
ArrDelay             873
DepDelay               0
Origin                 0
Dest                   0
Distance               0
TaxiIn               733
TaxiOut               44
Cancelled              0
dtype: int64


### Luego de utilizar el metodo isna() y sumarlos con sum() podemos ver que tenemos datos faltantes, en las columnas , ArrTime , TailNum, ActualElapsedTime, CRSElapsedTime , AirTime, ArrDelay ,  TaxiIn y TaxiOut  ya que estos datos no pueden ser abordados desde una media o un tiempo estimado ademas del data set ser lo suficientemente grande como para poder obtener buenas metricas procederemos a eliminar las filas que contengan registros faltantes.

In [74]:
my_df_cleaned = my_df_sampled.dropna(subset=['ArrTime' , 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime' , 'AirTime', 'ArrDelay' ,  'TaxiIn', 'TaxiOut'])
my_df_cleaned.head(21)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1782417     12           6          6   1805.0        1755   2146.0   
512712       3          23          7   2120.0        1200   2241.0   
447137       3          31          1   1731.0        1710   1819.0   
55082        1           6          7   1507.0        1413   1601.0   
877634       5          22          4    637.0         630    812.0   
1056704      6          12          4   2136.0        2125   2300.0   
198754       2          14          4   1807.0        1800   1910.0   
532837       3          23          7   1315.0        1305   1937.0   
1293928      8          31          7   1550.0        1530   1704.0   
749884       5          22          4   1814.0        1640   1923.0   
563348       3          18          2   2128.0        1805     42.0   
1514241      9          30          2   1218.0        1200   1416.0   
256840       2           9          6   1202.0        1126   1348.0   
198813       2          14          4   2123.0        2040   2312.0   
1171934      7           2          3   2240.0        2130   2353.0   
212103       2          23          6   1919.0        1900   2040.0   
812083       5           2          5   1512.0        1505   1710.0   
719854       4          15          2   1525.0        1420   1834.0   
1252870      7          17          4   1332.0        1316   1633.0   
1820361     12          15          1   1841.0        1811   2046.0   
841626       5           2          5   1236.0        1225   1429.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1782417        2035            YV       7281  N505MJ              281.0   
512712         1333            NW        131  N777NC               81.0   
447137         1810            OO       3676  N224SW               48.0   
55082          1509            OO       6573  N708SK               54.0   
877634          750            CO       1869  N16646              155.0   
1056704        2230            AA       2479  N3BHAA              204.0   
198754         1900            WN         47  N523SW               63.0   
532837         1915            AA        681  N5DYAA              262.0   
1293928        1652            XE       7776  N10575               74.0   
749884         1755            WN       1412  N468WN               69.0   
563348         2122            CO       1408  N37287              134.0   
1514241        1405            AS        326  N551AS              118.0   
256840         1319            UA        339  N812UA              166.0   
198813         2235            WN        534  N606SW              109.0   
1171934        2305            WN       3758  N658SW              133.0   
212103         2025            WN        456  N239WN               81.0   
812083         1648            EV       4316  N870AS              118.0   
719854         1720            CO        377  N24212              189.0   
1252870        1623            CO       1083  N27205              241.0   
1820361        2018            UA       1505  N467UA              125.0   
841626         1410            9E       2039  91479E              113.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1782417           220.0    177.0      71.0      10.0    IAD  AUS      1297   
512712             93.0     55.0     548.0     560.0    ORD  MSP       334   
447137             60.0     44.0       9.0      21.0    SLC  PIH       150   
55082              56.0     32.0      52.0      54.0    HDN  DEN       142   
877634            140.0    134.0      22.0       7.0    JAX  IAH       817   
1056704           185.0    174.0      30.0      11.0    DFW  LAX      1235   
198754             60.0     47.0      10.0       7.0    DAL  HOU       239   
532837            250.0    216.0      22.0      10.0    LAX  ORD      1745   
1293928            82.0     60.0      12.0      20.0    LAX  SMF       373   
749884         

In [75]:
check_datos_faltantes = my_df_cleaned.isna().sum()
check_datos_faltantes

Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
dtype: int64

### De esta manera comprobamos que no hayan datos faltantes ya que sea han eliminado los registros faltantes con el metodo .dropna() 

### A continuacion comprabaremos si hay valores duplicados dentro de nuestro DataSet 

In [76]:
df_cleaned = my_df_cleaned.duplicated()
sum_duplicates = df_cleaned.sum()
print("Cantidad de filas duplicadas:", sum_duplicates)

Cantidad de filas duplicadas: 0


## Crea columnes noves (velocitat mitjana del vol)


In [77]:
my_df_cleaned['VelocidadMediaMph'] = my_df_cleaned['Distance'] / (my_df_cleaned['AirTime'] / 60)
my_df_cleaned.head()

/var/folders/qn/t_d0_whn09sdmbx1px99544c0000gn/T/ipykernel_874/286609310.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  my_df_cleaned['VelocidadMediaMph'] = my_df_cleaned['Distance'] / (my_df_cleaned['AirTime'] / 60)


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1782417     12           6          6   1805.0        1755   2146.0   
512712       3          23          7   2120.0        1200   2241.0   
447137       3          31          1   1731.0        1710   1819.0   
55082        1           6          7   1507.0        1413   1601.0   
877634       5          22          4    637.0         630    812.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1782417        2035            YV       7281  N505MJ              281.0   
512712         1333            NW        131  N777NC               81.0   
447137         1810            OO       3676  N224SW               48.0   
55082          1509            OO       6573  N708SK               54.0   
877634          750            CO       1869  N16646              155.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1782417           220.0    177.0      71.0      10.0    IAD  AUS      1297   
512712             93.0     55.0     548.0     560.0    ORD  MSP       334   
447137             60.0     44.0       9.0      21.0    SLC  PIH       150   
55082              56.0     32.0      52.0      54.0    HDN  DEN       142   
877634            140.0    134.0      22.0       7.0    JAX  IAH       817   

         TaxiIn  TaxiOut  Cancelled  VelocidadMediaMph  
1782417     5.0     99.0          0         439.661017  
512712      7.0     19.0          0         364.363636  
447137      3.0      1.0          0         204.545455  
55082       5.0     17.0          0         266.250000  
877634      5.0     16.0          0         365.820896

### Se ha creado una nueva columna en nuestro DataSet en donde el resultado de la division entre la Distancia (expresada en millas) y el Tiempo de Vuelo expresado en minutos dividido 60, no da como resultado la velocidad media en millas por hora.

## (si ha arribat tard o no) 

In [78]:
my_df_cleaned['Arrivo_Tardio'] = my_df_cleaned['ArrDelay'] >= 15
my_df_cleaned.head()

/var/folders/qn/t_d0_whn09sdmbx1px99544c0000gn/T/ipykernel_874/3925083278.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  my_df_cleaned['Arrivo_Tardio'] = my_df_cleaned['ArrDelay'] >= 15


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1782417     12           6          6   1805.0        1755   2146.0   
512712       3          23          7   2120.0        1200   2241.0   
447137       3          31          1   1731.0        1710   1819.0   
55082        1           6          7   1507.0        1413   1601.0   
877634       5          22          4    637.0         630    812.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1782417        2035            YV       7281  N505MJ              281.0   
512712         1333            NW        131  N777NC               81.0   
447137         1810            OO       3676  N224SW               48.0   
55082          1509            OO       6573  N708SK               54.0   
877634          750            CO       1869  N16646              155.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1782417           220.0    177.0      71.0      10.0    IAD  AUS      1297   
512712             93.0     55.0     548.0     560.0    ORD  MSP       334   
447137             60.0     44.0       9.0      21.0    SLC  PIH       150   
55082              56.0     32.0      52.0      54.0    HDN  DEN       142   
877634            140.0    134.0      22.0       7.0    JAX  IAH       817   

         TaxiIn  TaxiOut  Cancelled  VelocidadMediaMph  Arrivo_Tardio  
1782417     5.0     99.0          0         439.661017           True  
512712      7.0     19.0          0         364.363636           True  
447137      3.0      1.0          0         204.545455          False  
55082       5.0     17.0          0         266.250000           True  
877634      5.0     16.0          0         365.820896           True

### Se ha creado una columna llamda Arrivo Tardio a partir del calculo de la columna que nos trae el DataSet  (ArrDelay)  en donde se informa de la llega tarde o no de los vuelos, en este caso si el tiempo es mayor a 15 minutos se considera tarde.

## Fes una taula de les aerolínies amb més endarreriments acumulats.

In [79]:
airline_total_delays = my_df_cleaned.groupby('UniqueCarrier')['ArrDelay'].sum()
airline_total_delays = airline_total_delays.sort_values(ascending=False)
airline_total_delays.head(10)

UniqueCarrier
WN    1178394.0
AA     910681.0
UA     696819.0
MQ     665562.0
OO     619404.0
XE     543941.0
DL     470912.0
CO     427113.0
EV     402521.0
YV     380899.0
Name: ArrDelay, dtype: float64

### Teniendo como columna UniqueCarrier (nombres de las aerolineas) y ArrDelay (con la cantidad de minutos de retraso acumulados) podemos agrupar las aerolineas con el metodo groupby() y sumar lo minutos con el metodo .sum() para luego ordenar los valores de forma descendiente y asi poder generar un nuevo cuadro con los datos obtenidos. 

## Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

In [80]:
VuelosMasLargos = my_df_cleaned.nlargest(200000, 'Distance')[['FlightNum', 'Distance']]

mas_largos_no_duplicados = VuelosMasLargos.drop_duplicates()

mas_largos_no_duplicados.head(10)


FlightNum  Distance
1622562         15      4962
1066841         14      4962
281487        1282      4502
567490         850      4502
1428660       1273      4502
1259672       1560      4502
569476         851      4502
1577757       1561      4502
1323548          1      4243
1036520         72      4243

### Estos fueron los vuelos mas largos sin repetir el numero de vuelo.

In [81]:
VuelosMasRetrasados = my_df_cleaned.nlargest(10, 'ArrDelay')[['FlightNum', 'Distance', 'ArrDelay']]
mas_retrasados_no_duplicados = VuelosMasRetrasados.drop_duplicates()
mas_retrasados_no_duplicados.head(10)

FlightNum  Distance  ArrDelay
686014        1699       930    2453.0
685437        1472       449    1542.0
521096        1743       200    1490.0
1400101       1968       988    1286.0
1502764        282       813    1186.0
1193616        927       665    1175.0
1213016        435      1307    1158.0
66969           47      2486    1081.0
1021378       4829       812    1057.0
1898129        638       813    1049.0

### Estos fueron los 10 vuelos mas retrasados.

## Aporta allò que consideris rellevant.

### Luego de rehacer el ejercicio por completo, podemos ver que es de suma importancia interpretar bien cada variable del DataSet para poder tomar las deciciones adecuadas con respecto a los ejercicios planteados. 
### Ademas de conocer sobre que tratan las variables la preaparacion de los datos al principio es un momento clave para luego poder operar con mayor facilidad, entre ello luego de cargar los datos,  es importante revisar la estructura de los datos, verificar el tamaño y operar con una cantidad que el ordenador pueda procesar, hay que saber que hacer tanto con los valores NaN como los duplicados.

# - Exercici 3
## Exporta el dataset net i amb les noves columnes a Excel.

In [ ]:
my_df_cleaned.to_excel('MyDatasetAirLines.xlsx')